src/relay.cpp

```diff
@@ -961,50 +932,52 @@ void loop_relay(std::vector<relay_config> *vlans) {
             state_db.get(), "HW_MUX_CABLE_TABLE"
         );
 
-    int filter = 0;
-    filter = sock_open(&ether_relay_fprog);
-    sockets.push_back(filter);
+    auto filter = sock_open(&ether_relay_fprog);
+    if (filter != -1) {
+        sockets.push_back(filter);
+        listen_event = event_new(base, filter, EV_READ|EV_PERSIST, client_callback,
+                                 reinterpret_cast<void *>(&vlans));
+        if (listen_event == NULL) {
+            syslog(LOG_ERR, "libevent: Failed to create client listen libevent\n");
+        }
+        event_add(listen_event, NULL);
+        syslog(LOG_INFO, "libevent: add filter socket event\n");
+    }
 
-    for(relay_config &vlan : *vlans) {
-        relay_config *config = &vlan;
-        int local_sock = 0; 
+    for(auto &vlan : vlans) {
+        int local_sock = 0;
         int server_sock = 0;
-        config->config_db = config_db;
-        config->mux_table = mStateDbMuxTablePtr;
-        config->state_db = state_db;
-        config->mux_key = vlan_member + config->interface + "|";
+        vlan.second.config_db = config_db;
+        vlan.second.mux_table = mStateDbMuxTablePtr;
+        vlan.second.state_db = state_db;
+        vlan.second.mux_key = vlan_member + vlan.second.interface + "|";
+
+        update_vlan_mapping(vlan.first, config_db);
 
         std::string counterVlan = counter_table;
-        initialize_counter(config->state_db, counterVlan.append(config->interface));
-        prepare_socket(&local_sock, &server_sock, config);
+        initialize_counter(vlan.second.state_db, counterVlan.append(vlan.second.interface));
+        prepare_socket(local_sock, server_sock, vlan.second);
 
-        config->local_sock = local_sock;
-        config->server_sock = server_sock;
+        vlan.second.local_sock = local_sock;
+        vlan.second.server_sock = server_sock;
 
         sockets.push_back(local_sock);
         sockets.push_back(server_sock);
-        prepare_relay_config(config, &local_sock, filter);
+        prepare_relay_config(vlan.second, local_sock, filter);
 
         evutil_make_listen_socket_reuseable(filter);
         evutil_make_socket_nonblocking(filter);
 
         evutil_make_listen_socket_reuseable(local_sock);
         evutil_make_socket_nonblocking(local_sock);
     
-        if (dual_tor_sock) {
-            listen_event = event_new(base, filter, EV_READ|EV_PERSIST, callback_dual_tor, config);
-        }
-        else {
-            listen_event = event_new(base, filter, EV_READ|EV_PERSIST, callback, config);
-        }
-	server_listen_event = event_new(base, local_sock, EV_READ|EV_PERSIST, server_callback, config);
+	    server_listen_event = event_new(base, local_sock, EV_READ|EV_PERSIST, server_callback, &(vlan.second));
```

Comment:
I use unordered _map since updating map will not change value memory. And we don't support vlan dynamic add/delete. But I understand your concern, since once registered this event, callback will always use the address. Maybe it's better to pass vlan interface name as argument, and search vlan config in callback handler. 
I will think about it and find a better solution.

Thank you again for this good comment.

Reply:
I think even if we insert/remove elements, it doesn't invalidate the references of the values in the map. So, it's okay to have this approach.